### PR TITLE
linuxPackages.ati_drivers_x11: patch for kernel 4.7+

### PR DESCRIPTION
**Pull request by jerith666**
###### Motivation for this change

Yet another case of binary ATI drivers lagging behind kernel API changes.
###### Things done
- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [x] NixOS
  - [ ] OS X
  - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

uses patch from commit 392aac08 in https://aur.archlinux.org/packages/catalyst/

Confirmed working with minecraft and starcraft 2 locally.

### CODE REVIEW

File: pkgs/os-specific/linux/ati-drivers/default.nix [hunk 1]
```diff
@@ -77,7 +77,10 @@ stdenv.mkDerivation rec {
   ++ optionals ( kernel != null &&
                  (builtins.compareVersions kernel.version "4.6") >= 0 )
                [ ./patches/kernel-4.6-get_user_pages.patch
-                 ./patches/kernel-4.6-page_cache_release-put_page.patch ];
+                 ./patches/kernel-4.6-page_cache_release-put_page.patch ]
+  ++ optionals ( kernel != null &&
+                 (builtins.compareVersions kernel.version "4.7") >= 0 )
```
Comment (NeQuissimus) [inline pkgs/os-specific/linux/ati-drivers/default.nix]: Could we not use `versionAtLeast` and/or `versionOlder` instead? Seems cleaner

Comment (jerith666) [inline pkgs/os-specific/linux/ati-drivers/default.nix]: I thought I looked for something like that when I did the 4.6 patches and didn't find it.  I'll try to revise tonight.

**Comment by grahamc** [comment]
>  Yet another case of binary ATI drivers lagging behind kernel API changes.

Hey, Matt!

Thank you so much for the contribution, and keeping these up to date. I know how frustrating it is to find out after the fact that kernel updates have broken your drivers, I've had the same experience with the nvidia drivers myself.

If you would like, you could add yourself as a maintainer to the package, and when we get a kernel update you'll get an email when the package fails. Then you woudl know before you try and upgrade, and give you a chance to patch it sooner. This would be a huge help if you're willing.

Thank you for the patch!

**Comment by RonnyPfannschmidt** [comment]
its very unfortunate, but i fear #19841 makes this one invalid

**Comment by jerith666** [comment]
I think kernel 4.8, 4.9, etc. will maintain the same API changes that broken with the 4.7 release, though, no?  We are doing `versionAtLeast`, after all.  :)  (The thing that might make this invalid would be a new `ati-drivers` release.)

That said, I haven't tested with kernel 4.8.  I can do that if you'd like before merging.

**Comment by grahamc** [comment]
Hey @jerith666, sorry for being behind on this. If you could test it with 4.8 (if you'd go so far as building your system and booting to it, that would be ideal) I'd merge it ASAP.

**Comment by jerith666** [comment]
Sorry, I'm behind on it too!  I actually did pick it back up last night and got so far as to build it successfully with kernel 4.8, but I haven't booted it yet.  Hoping to do that tonight (while waiting for election returns to roll in ... :smile: ).

**Comment by jerith666** [comment]
I can now confirm that this works with kernel 4.8.4 in recent `nixos-unstable` @ fa4167c.  StarCraft II & minecraft are both working.

**Pull request merged**